ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadState.java

```diff
@@ -0,0 +1,218 @@
+    this.maxOutstandingRequests = maxOutstandingRequests;
+    this.future = new CompletableFuture<>();
+    this.tearDownFuture = new CompletableFuture<>();
+    this.future.whenComplete(this::doCancel);
+    this.tearDownFuture.exceptionally(
+        error -> {
+          // Propagate cancellation back to our internal future.
+          if (error instanceof CancellationException) {
+            this.future.cancel(true);
+          }
+          return null;
+        });
+  }
+
+  private synchronized void doCancel(final Void result, final Throwable error) {
+    // Handle cancellations
+    if (future.isCancelled()) {
+      LOG.info("World state download cancelled");
+    } else if (error != null) {
+      if (!(ExceptionUtils.rootCause(error) instanceof StalledDownloadException)) {
+        LOG.info("World state download failed. ", error);
+      }
+    }
+    if (persistenceTask != null) {
+      persistenceTask.cancel();
+    }
+    for (final EthTask<?> outstandingRequest : outstandingRequests) {
+      outstandingRequest.cancel();
+    }
+    pendingRequests.clear();
+    requestsToPersist.clear();
+    if (error != null) {
+      tearDownFuture.completeExceptionally(error);
+    } else {
+      tearDownFuture.complete(result);
+    }
+  }
+
+  public void whileAdditionalRequestCanBeSent(final Runnable action) {
+    while (shouldRequestNodeData()) {
+      if (sendingRequests.compareAndSet(false, true)) {
+        try {
+          action.run();
+        } finally {
+          sendingRequests.set(false);
+        }
+      } else {
+        break;
+      }
+    }
+  }
+
+  public synchronized void setWaitingForNewPeer(final boolean waitingForNewPeer) {
+    this.waitingForNewPeer = waitingForNewPeer;
+  }
+
+  public synchronized void addOutstandingTask(final EthTask<?> task) {
+    outstandingRequests.add(task);
+  }
+
+  public synchronized void removeOutstandingTask(final EthTask<?> task) {
+    outstandingRequests.remove(task);
+  }
+
+  public int getOutstandingRequestCount() {
+    return outstandingRequests.size();
+  }
+
+  private synchronized boolean shouldRequestNodeData() {
+    return !future.isDone()
+        && outstandingRequests.size() < maxOutstandingRequests
+        && !pendingRequests.isEmpty()
+        && !waitingForNewPeer;
+  }
+
+  public CompletableFuture<Void> getFuture() {
+    return tearDownFuture;
+  }
+
+  public synchronized void setPersistenceTask(final EthTask<?> persistenceTask) {
+    this.persistenceTask = persistenceTask;
+  }
+
+  public synchronized void enqueueRequest(final NodeDataRequest request) {
+    if (!future.isDone()) {
+      pendingRequests.enqueue(request);
+    }
+  }
+
+  public synchronized void enqueueRequests(final Stream<NodeDataRequest> requests) {
+    if (!future.isDone()) {
+      requests.forEach(pendingRequests::enqueue);
+    }
+  }
+
+  public synchronized void markAsFailed(final Task<NodeDataRequest> task) {
+    if (!future.isDone()) {
+      task.markFailed();
```

Comment:
just an fyi - if `task.markFailed()` is called after the queue is cleared, it doesn't requeue the failed task.

Reply:
hmm, that's true. I know I had a test failing when I added it but it definitely seems fine now and all the code checks out. It's much nicer without it so removing...